src/dfx/src/lib/nns/install_nns.rs

```diff
@@ -1,32 +1,661 @@
+/// - Returns an error if the replica URL could not be found.  Typically this indicates that the local replica
+///   is not running or is running in a different location.
+///
+/// # Panics
+/// This code is not expected to panic.
+pub fn get_replica_url(env: &dyn Environment) -> anyhow::Result<Url> {
+    let port = get_running_replica_port(
+        env,
+        env.get_network_descriptor()
+            .local_server_descriptor
+            .as_ref()
+            .ok_or_else(|| anyhow!("Could not get the local server descriptor"))?,
+    )?
+    .ok_or_else(|| anyhow!("Could not determine the port of the local server"))?;
+    let url = Url::parse(&format!("http://localhost:{port}"))?;
+    Ok(url)
+}
+
+/// Arguments for the ic-nns-init command line function.
+pub struct IcNnsInitOpts {
+    /// An URL to accees one or more NNS subnet replicas.
+    nns_url: String,
+    /// A directory that needs to be populated will all required wasms before calling ic-nns-init.
+    wasm_dir: PathBuf,
+    /// The ID of a test account that ic-nns-init will create and to initialise with tokens.
+    test_accounts: Option<String>, // TODO, does the CLI actually support several?
+    /// A subnet for SNS canisters.
+    sns_subnets: Option<String>, // TODO: Can there be several?
+}
+
+/// Calls the `ic-nns-init` executable.
+///
+/// Notes:
+///   - Set DFX_IC_NNS_INIT_PATH=<path to binary> to use a different binary for local development
+///   - This won't work with an HSM, because the agent holds a session open
+///   - The provider_url is what the agent connects to, and forwards to the replica.
+#[context("Failed to install nns components.")]
+pub async fn ic_nns_init(ic_nns_init_path: &Path, opts: &IcNnsInitOpts) -> anyhow::Result<()> {
     let mut cmd = std::process::Command::new(ic_nns_init_path);
-    cmd.arg("--help");
+    cmd.arg("--url");
+    cmd.arg(&opts.nns_url);
+    cmd.arg("--wasm-dir");
+    cmd.arg(&opts.wasm_dir);
+    opts.test_accounts.iter().for_each(|account| {
+        cmd.arg("--initialize-ledger-with-test-accounts");
+        cmd.arg(account);
+    });
+    opts.sns_subnets.iter().for_each(|subnet| {
+        cmd.arg("--sns-subnet");
+        cmd.arg(subnet);
+    });
     cmd.stdout(std::process::Stdio::inherit());
     cmd.stderr(std::process::Stdio::inherit());
     let output = cmd
         .output()
         .with_context(|| format!("Error executing {:#?}", cmd))?;
 
     if !output.status.success() {
-        bail!("ic-nns-init call failed");
+        return Err(anyhow!("ic-nns-init call failed"));
+    }
+    Ok(())
+}
+
+/// Sets the exchange rate between ICP and cycles.
+///
+/// # Implementation
+/// This is done by proposal.  Just after startung a test server, ic-admin
+/// proposals with a test user pass immediately, as the small test neuron is
+/// the only neuron and has absolute majority.
+pub fn set_xdr_rate(rate: u64, nns_url: &Url) -> anyhow::Result<()> {
+    std::process::Command::new("ic-admin")
```

Comment:
`ic-admin` is available in the binary cache.  You can get it with env.get_cache().get_binary_command_path("ic-admin")?;